File: Ex0508_IsLegalStackOperation/IsLegalOper.java
import java.util.Scanner;

public class IsLegalOper {
    public static void main(String[] args) {
        int n, m;
        Scanner scanner = new Scanner(System.in);

        n = scanner.nextInt();
        m = scanner.nextInt();


        // Clean the buffer
        scanner.nextLine();

        String[] strings = new String[n];
        for (int i = 0; i < n; i++) {
            strings[i] = scanner.nextLine();
        }

        for (String s: strings) {
            if (isLegal(s, m)) {
                System.out.println("YES");
            }
            else {
                System.out.println("NO");
            }
        }
    }

    public static boolean isLegal(final String s, int m) {
        int currentSize = 0;
        for (int i = 0; i < s.length(); i++) {
            switch (s.charAt(i)) {
                case 'S':
                    currentSize++;
                    break;
                case 'X':
                    currentSize--;
                    break;
                default:
                    return false;
            }
            if (currentSize < 0 || currentSize > m) {
                return false;
            }
        }

        return currentSize == 0;
    }
}

File: Ex0508_IsLegalStackOperation/IsLegalOper.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool IsLegal(string S, int M)
{
	int current_size = 0;
	for (int i = 0; i < (int)S.length(); i++)
	{
		switch (S[i])
		{
		case 'S':
			current_size++;
			break;
		case 'X':
			current_size--;
			break;
		default:
			return false;
			break;
		}
		if (current_size<0 || current_size>M)
		{
			return false;
		}
	}
	if (current_size == 0)
	{
		return true;
	}
	else
	{
		return false;
	}
}

int main(void)
{
	int N = 0, M = 50;
	cin >> N >> M;

	// Clean the buffer
	cin.ignore(1024, '\n');

	vector<string> Strings;
	for (int i = 0; i < N; i++)
	{
		string temp;
		getline(cin, temp);
		Strings.push_back(temp);
	}
	for (int i = 0; i < (int)Strings.size(); i++)
	{
		if (IsLegal(Strings[i], M))
		{
			cout << "YES" << endl;
		}
		else
		{
			cout << "NO" << endl;
		}
	}
	
	return 0;
}
